File: 0368-largest-divisible-subset/0368-largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int len = nums.size();
        sort(nums.begin(), nums.end());

        vector<int> res;
        int mxSize = 0;
        vector<vector<int>> dp(len, vector<int>(0));
        for (int i = 0; i < len; i++) {
            vector<int> temp = solve(i, nums, dp);
            int currSize = temp.size();
            if (currSize > mxSize) {
                mxSize = currSize;
                res = temp;
            }
        }

        return res;        
    }

private:
    vector<int> solve(int n, vector<int> &nums, vector<vector<int>> &dp) {
        if (n < 0) {
            return {};
        }

        if (dp[n].size() != 0) {
            return dp[n];
        }

        vector<int> res = {nums[n]};
        int mxSize = 1;
        for (int i = 0; i < n; i++) {
            if (nums[n] % nums[i] == 0) {
                vector<int> temp = solve(i, nums, dp);
                temp.push_back(nums[n]);

                int currSize = temp.size();
                if (currSize > mxSize) {
                    mxSize = currSize;
                    res = temp;
                }
            }
        }

        return dp[n] = res;
    }
};

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int len = nums.length;
        Arrays.sort(nums);
        int[] lis = new int[len];
        int[] prevIdx = new int[len];
        for (int i = 0; i < len; i++) {
            lis[i] = 1;
            prevIdx[i] = i;
        }

        int globalMax = 1;
        int globalMaxIdx = 0;
        for (int curr = 0; curr < len; curr++) {
            int currMax = 1;
            int maxIdx = curr;
            for (int prev = 0; prev < curr; prev++) {
                if ((nums[curr] % nums[prev]) == 0 && lis[prev] + 1 > currMax) {
                    currMax = lis[prev] + 1;
                    maxIdx = prev;
                }
            }

            lis[curr] = currMax;
            prevIdx[curr] = maxIdx;

            if (currMax > globalMax) {
                globalMax = currMax;
                globalMaxIdx = curr;
            }
        }

        // System.out.println(Arrays.toString(lis));
        // System.out.println(Arrays.toString(prevIdx));

        List<Integer> ans = new ArrayList<>();
        int temp = globalMaxIdx;
        for (int i = 0; i < globalMax; i++) {
            ans.add(nums[temp]);
            temp = prevIdx[temp];
        }

        return ans;
    }
}
